backend/src/routes/api/users.ts

```diff
@@ -0,0 +1,64 @@
+import express from "express";
+import bcryptjs from "bcryptjs";
+import jsonwebtoken from "jsonwebtoken";
+import User from "../../models/user";
+
+const router = express.Router();
+
+//@route    Post api/users
+//@desc     Register new user
+//@access   Public
+router.post("/", (req, res) => {
+  const { name, email, password, userType } = req.body;
+
+  //Simple validation
+  if (!name || !email || !password || !userType) {
+    return res.status(400).json({ msg: "Please enter all fields" });
+  }
+
+  if (!["tenant", "owner", "realEstate"].includes(userType)) {
+    return res.status(400).json({ msg: "Invalid user type" });
+  }
+
+  //Check for existing user
+  User.findOne({ email: email }).then((user) => {
+    if (user) return res.status(400).json({ msg: "User Already exists" });
+
+    const newUser = new User({
+      name,
+      email,
+      password,
+      userType,
+    });
+
+    //Create salt & hash
+    bcryptjs.genSalt(10, (err, salt) => {
+      bcryptjs.hash(newUser.password, salt, (err, hash) => {
+        if (err) throw err;
+        newUser.password = hash;
+        newUser.save().then((user) => {
+          //sign token
+          jsonwebtoken.sign(
+            { id: user.id },
+            process.env.JWT_SECRET,
+            { expiresIn: 3600 },
+            (err, token) => {
+              if (err) throw err;
+              res.json({
+                token,
+                user: {
+                  id: user.id,
+                  name: user.name,
+                  email: user.email,
+                  type: user.userType,
+                },
+              });
+            }
+          );
+        });
+      });
+    });
+  });
```

Comment:
Particularmente no soy fan de los anidamientos, me cuesta leerlos.
Te dejo una suggestion de como cambiaria este pedazote de código, obvio es a gusto, si no te parece, descartalo y fue.

```suggestion
  try {
    const salt: string = await bcryptjs.genSalt(10);
    const hash: string = await bcryptjs.hash(newUser.password, salt); // btw: you can just send the amount of "salts" in here and not generate them with salt.
    newUser.password = hash;

    const user = await newUser.save();
    const token = jsonwebtoken.sign({id: user.id}, process.env.JWT_SECRET, {expiresIn: 3600});

    if (!token) {
      res.status(400).json({msg: 'Error generating token'});
    }
    res.json({
      token,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        type: user.userType,
      },
    });
  } catch (error) {
    res.status(400).json({msg: `Error registering user: ${error}`});
  }
```